### ISSUE TITLE
For `RefCell<T> where T: Copy` suggest `Cell<T>` instead

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/rust-clippy#17380]
### What it does

In type contexts (or maybe only some of them like field definition?), if a `std::cell::RefCell` is found whose inner type implements `Copy`, suggest using `Cell` instead

### Advantage

- Remove memory footprint for the `usize` borrow counter inside `RefCell`
- Remove memory footprint for potential padding for the cell to be aligned like `usize`. Together with the above, `RefCell<bool>` is 16 bytes v.s. 1 byte for `Cell<bool>`
- Reduce CPU instructions for incrementing and decrementing a borrow counter when accessing the value

### Drawbacks

This is not a drop-in replacement. For example if some code (that is unpractical to modify) expects a `&mut T` value, `RefCell<T>` can provide it but `Cell<T>` cannot.

Maybe this lint should be allow by default?

### Example

```rust
pub struct Thing {
    ready: std::cell::RefCell<bool>,
}

impl Thing {
    pub fn thing(&self) {
        if !*self.ready.borrow() {
            self.do_the_thing()
            *self.ready.borrow_mut() = true;
        }
    }
}
```

Could be written as:

```rust
pub struct Thing {
    ready: std::cell::Cell<bool>,
}

impl Thing {
    pub fn thing(&self) {
        if !self.ready.get() {
            self.do_the_thing()
            self.ready.set(true);
        }
    }
}
```

### Comparison with existing lints

The example code triggers no existing lint with this template

**Comment by qdot3** [comment]
I'd like to implement this. The lint name will be `refcell_cell` after `mutex_atomic`.  The category will be `nursery` or `perf`.

**Comment by SimonSapin** [comment]
I wonder if `Copy` is the best criteria, since `Cell<T>` can be used even if `T: !Copy`. It’s only some methods like `get` that become unavailable.

Looking at the existing `mutex_atomic` and `mutex_integer`, I guess they match on a specific list of inner types? Although that’s because of corresponding `Atomic*` types

**Comment by qdot3** [comment]
IMO, the `Copy` bound is a reasonable criterion. As you pointed out, `Cell` can be used with `!Copy` types, but accessing the inner value requires it to be temporarily replaced with another value, which may be costly or less ergonomic. In such cases, `RefCell` can be a better fit because it provides borrowed access instead. However, this issue does not arise for `Copy` types.

`Atomic*` types can be enumerated explicitly because they are determined by available hardware support. In contrast, `Copy` types cannot be identified by listing concrete type names.

**Comment by qdot3** [comment]
The implementation is straightforward for definition sites. However, for type aliases, local expressions, and local variables, whether `Cell` or `RefCell` is preferable depends on how the value is used. For example, `Cell` may be a better choice for a counter shared among functions within a thread, while some function APIs may require `RefCell` due to their interface constraints. A fully accurate analysis of these cases would require fine-grained usage analysis and significantly increase the complexity of the lint. Therefore, I am currently considering leaving them out.

I would appreciate feedback on whether this trade-off seems reasonable, or whether these cases should be handled despite the additional complexity.

**Comment by SimonSapin** [comment]
> some function APIs may require `RefCell` due to their interface constraints

I think this is what I meant by “if some code (that is unpractical to modify) expects a `&mut T` value”. I agree it’s not practical to mechanically detect such cases, which is why I suggested the lint be `allow` (silent/disabled) by default.

I don’t know how this relates to (type?) definition sites v.s. type aliases, local expressions, and local variables.

**Comment by qdot3** [comment]
> I agree it’s not practical to mechanically detect such cases, which is why I suggested the lint be allow (silent/disabled) by default.

Oh, I see.

> I don’t know how this relates to (type?) definition sites v.s. type aliases, local expressions, and local variables.

The latter requires trait‑bounds checks and other usage‑based analysis; otherwise this lint can become very noisy. This isn’t an issue for the former, because a field or function definition site is checked only once.
